crates/functional/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+                state.o3 = o3.clone();
+                state.o4 = o4.clone();
+                state.o5 = o5.clone();
+                state.destructor.replace(Box::new(new_destructor));
+            } else if state.destructor.is_none() {
+                should_update = true;
+                state.destructor.replace(Box::new(callback(
+                    state.o1.borrow(),
+                    state.o2.borrow(),
+                    state.o3.borrow(),
+                    state.o4.borrow(),
+                    state.o5.borrow(),
+                )));
+            }
+            return move || {
+                if should_update {
+                    hook_update(move |_: &mut UseEffectState<T1, T2, T3, T4, T5, Destructor>| true)
+                }
+            };
+        },
+        || UseEffectState {
+            o1: o1_c,
+            o2: o2_c,
+            o3: o3_c,
+            o4: o4_c,
+            o5: o5_c,
+            destructor: None,
+        },
+    )();
+}
+
+pub fn use_hook<InternalHookState, HookRunner, R, InitialStateProvider, PretriggerChange: 'static>(
+    hook_runner: HookRunner,
+    initial_state_producer: InitialStateProvider,
+) -> R
+where
+    HookRunner: FnOnce(&mut InternalHookState, Box<dyn Fn(PretriggerChange)>) -> R,
+    InternalHookState: 'static,
+    InitialStateProvider: FnOnce() -> InternalHookState,
+    PretriggerChange: FnOnce(&mut InternalHookState) -> bool,
+{
+    // Extract current hook
+    let (mut component_hook_state, hook, render_call) = CURRENT_HOOK.with(|hook_state_holder| {
+        let component_hook_state = hook_state_holder.clone();
+        let hook_state_holder = hook_state_holder.try_borrow_mut();
+        let mut hook_state_holder = hook_state_holder.expect("Nested hooks not supported");
+        let mut hook_state = hook_state_holder
+            .as_mut()
+            .expect("No current hook. Hooks can only be called inside functional components");
+
+        // Determine which hook position we're at and increment for the next hook
+        let hook_pos = hook_state.counter;
+        hook_state.counter += 1;
+
+        // Initialize hook if this is the first call
+        if hook_pos >= hook_state.hooks.len() {
+            let initial_state = Rc::new(RefCell::new(initial_state_producer()));
+            hook_state.hooks.push(initial_state);
+        }
+
+        let hook = hook_state.hooks[hook_pos].clone();
+        let render_call = hook_state.render_call.clone();
+
+        return (component_hook_state, hook, render_call);
+    });
+
+    let trigger = {
+        let hook = hook.clone();
+        Box::new(move |pretrigger_change: PretriggerChange| {
+            let mut hook = hook.clone();
+            component_hook_state
+                .try_borrow_mut()
+                .expect("Hook is being modified elsewhere. Note that hooks cannot be nested.")
+                .deref_mut()
+                .as_mut()
+                .unwrap()
+                .next_tick
```

Comment:
This doesn't look right. Change triggers should be immediate. No need to optimize at this level, yew framework should be responsible for making sure only one render happens for multiple changes.

Reply:
Refactored